schema-engine/connectors/sql-schema-connector/src/flavour/postgres.rs

```diff
@@ -462,8 +467,33 @@ impl SqlFlavour for PostgresFlavour {
                     let ret =
                         shadow_db::sql_schema_from_migrations_history(migrations, shadow_database, namespaces).await;
 
-                    let drop_database = format!("DROP DATABASE IF EXISTS \"{shadow_database_name}\"");
-                    main_connection.raw_cmd(&drop_database, &params.url).await?;
+                    if is_postgres {
+                        // When drop database is routed through pgbouncer, the database may still be used in other pooled connections.
+                        // In this case, given that we (as a user) know the database will not be used any more, we can forcefully drop
+                        // the database. Note that `with (force)` is added in Postgres 13, and therefore we will need to
+                        // fallback to the normal drop if it errors with syntax error.
+                        //
+                        // TL;DR,
+                        // 1. pg >= 13 -> it works.
+                        // 2. pg < 13 -> syntax error on WITH (FORCE), and then fail with db in use.
+                        let drop_database = format!("DROP DATABASE IF EXISTS \"{shadow_database_name}\" WITH (FORCE)");
+                        if let Err(err) = main_connection.raw_cmd(&drop_database, &params.url).await {
+                            if let Some(msg) = err.message() {
+                                if msg.contains("syntax error") {
+                                    let drop_database_alt =
+                                        format!("DROP DATABASE IF EXISTS \"{shadow_database_name}\"");
+                                    main_connection.raw_cmd(&drop_database_alt, &params.url).await?;
+                                } else {
+                                    return Err(err);
+                                }
+                            } else {
+                                return Err(err);
+                            }
+                        }
+                    } else {
+                        let drop_database = format!("DROP DATABASE IF EXISTS \"{shadow_database_name}\"");
+                        main_connection.raw_cmd(&drop_database, &params.url).await?;
+                    }
```

Comment:
Okay, cool, I read back through the convo folks had internally and I think I agree with us just trying to use the new syntax and then defaulting to without `WITH (FORCE)` if it errs.

Nit on my side:
With the new chunk of logic and I’m somewhat of the mind of extracting it out into a new `drop_db` fn or some-such to keep things legible as I found it otherwise a little difficult to keep track of things with the amount of nesting.

But generally LGTM ✨